coral-lib/src/structs/rewards_tree.rs

```diff
@@ -0,0 +1,99 @@
+use ethers::abi::Token;
+use ethers::prelude::*;
+use ethers::types::{Address, U256};
+use ethers::utils::keccak256;
+use serde::{Deserialize, Serialize};
+
+/// Reward data for a single node operator
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct NoOpReward {
+    /// Node operator address from which we calculate the total rewards
+    pub address: Address,
+    /// Total of the rewards for all the validators
+    pub total_rewards: U256,
+}
+
+/// Reward merkle tree data with hash, node address and rewards list l
+#[derive(Clone, Debug)]
+pub struct RewardValidatorMerkleData {
+    pub hash: [u8; 32],
+    pub node: Address,
+    pub reward: NoOpReward,
+}
+
+/// Helper to generate merkle tree leaf
+/// With this leaf model
+/// keccak256(abi.encode(noopAddress, startEpoch, endEpoch, total))
+/// we double hash to avoid any second preimage attack
+/// as explained there https://www.rareskills.io/post/merkle-tree-second-preimage-attack
+/// and to match smart contract function _buildMerkleProof
+pub fn generate_merkle_leaf(
+    address: Address,
+    start_epoch: U256,
+    end_epoch: U256,
+    total: U256,
+) -> [u8; 32] {
+    let calldata = abi::encode(&[
+        Token::Address(address),
+        Token::Uint(start_epoch),
+        Token::Uint(end_epoch),
+        Token::Uint(total),
+    ]);
+    let hash = keccak256(calldata);
+    keccak256(hash)
```

Comment:
So @kamiyaa , we sync with @bxmmm1 and realize it was not the same impl that was done in the contract function [_buildMerkleProof](https://github.com/PufferFinance/puffer-contracts/blob/71a544141dc0fb4c5cd03c305211aa4778125227/l2-contracts/test/unit/L2RewardManager.t.sol#L625) - they use double hash to prevent this kind of attack https://www.rareskills.io/post/merkle-tree-second-preimage-attack